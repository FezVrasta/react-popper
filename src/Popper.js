// @flow
import deepEqual from "deep-equal";
import * as React from 'react';
import { createPopper } from '@popperjs/core';
import {
  type State,
  type Placement,
  type PositioningStrategy,
  type Instance,
  type VirtualElement,
  type Modifier,
  type ModifierArguments,
} from '@popperjs/core/lib';
import type { Style } from 'typed-styles';
import { ManagerReferenceNodeContext } from './Manager';
import { unwrapArray, setRef, shallowEqual } from './utils';
import { type Ref } from "./RefTypes";

type ReferenceElement = VirtualElement | HTMLElement | null;
type StyleOffsets = { top: number, left: number };
type StylePosition = { position: 'absolute' | 'fixed' };
type Modifiers = Array<$Shape<Modifier<any>>>

export type PopperArrowProps = {
  ref: Ref,
  style: StyleOffsets & Style,
};
export type PopperChildrenProps = {|
  ref: Ref,
  style: StyleOffsets & StylePosition & Style,

  placement: Placement,
  isReferenceHidden: ?boolean,
  hasPopperEscaped: ?boolean,

  update: () => Promise<null | $Shape<State>>,
  arrowProps: PopperArrowProps,
|};
export type PopperChildren = PopperChildrenProps => React.Node;

export type PopperProps = {
  children: PopperChildren,
  innerRef?: Ref,
  modifiers?: Modifiers,
  placement?: Placement,
  strategy?: PositioningStrategy,
  referenceElement?: ReferenceElement,
  onFirstUpdate?: ($Shape<State>) => void
};

type PopperState = {
  placement: ?Placement,
  styles: ?{
    popper: Style,
    arrow: Style,
  },
  isReferenceHidden: ?boolean,
  hasPopperEscaped: ?boolean,
};

const initialPopperStyle = {
  position: 'absolute',
  top: 0,
  left: 0,
  opacity: 0,
  pointerEvents: 'none',
};

const initialArrowStyle = {};

export class InnerPopper extends React.Component<PopperProps, PopperState> {
  static defaultProps = {
    placement: 'bottom',
    strategy: 'absolute',
    modifiers: [],
    referenceElement: undefined,
  };

  state = {
    data: undefined,
    placement: undefined,
  };

  popperInstance: ?Instance;

  popperNode: ?HTMLElement = null;
  arrowNode: ?HTMLElement = null;

  setPopperNode = (popperNode: ?HTMLElement) => {
    if (!popperNode || this.popperNode === popperNode) return;

    setRef(this.props.innerRef, popperNode);
    this.popperNode = popperNode;

    this.updatePopperInstance();
  };

  setArrowNode = (arrowNode: ?HTMLElement) => {
    this.arrowNode = arrowNode;
  };

  updateStateModifier: Modifier<{||}> = {
    name: 'reactPopperState',
    enabled: true,
    phase: 'write',
    fn: ({ state }: ModifierArguments<{||}>) => {
      const { placement, styles, modifiersData } = state;
      let isReferenceHidden: boolean;
      let hasPopperEscaped: boolean;

      if (modifiersData.hide) {
        isReferenceHidden = modifiersData.hide.isReferenceHidden;
        hasPopperEscaped = modifiersData.hide.hasPopperEscaped;
      }

      this.setState({
        placement,
        styles,
        isReferenceHidden,
        hasPopperEscaped,
      });
    },
  };

  getOptions = () => ({
    placement: this.props.placement,
    strategy: this.props.strategy,
    modifiers: [
      ...this.props.modifiers,
      {
        name: 'arrow',
        enabled: !!this.arrowNode,
        options: {
          element: this.arrowNode,
        },
      },
      {
        name: 'applyStyles',
        enabled: false
      },
      this.updateStateModifier,
    ],
    onFirstUpdate: this.props.onFirstUpdate
  });

  getPopperStyle = () =>
    !this.popperNode || !this.state.styles
      ? initialPopperStyle
      : this.state.styles.popper;

  getArrowStyle = () =>
    !this.arrowNode || !this.state.styles
      ? initialArrowStyle
      : this.state.styles.arrow;

  destroyPopperInstance = () => {
    if (!this.popperInstance) return;

    this.popperInstance.destroy();
    this.popperInstance = null;
  };

  updatePopperInstance = () => {
    this.destroyPopperInstance();

    const { popperNode } = this;
    const { referenceElement } = this.props;

    if (!referenceElement || !popperNode) return;

    this.popperInstance = createPopper(
      referenceElement,
      popperNode,
      this.getOptions()
    );
  };

  update = () => {
    if (this.popperInstance) {
      return this.popperInstance.update();
    } else {
      return Promise.resolve(null);
    }
  };

  componentDidUpdate(prevProps: PopperProps, prevState: PopperState) {
    // If the Popper.js reference element has changed, update the instance (destroy + create)
    if (this.props.referenceElement !== prevProps.referenceElement) {
      this.updatePopperInstance();
    }

    // If the Popper.js options have changed, set options
    if (
      this.props.placement !== prevProps.placement ||
      this.props.strategy !== prevProps.strategy ||
      !deepEqual(this.props.modifiers, prevProps.modifiers, {strict: true})
    ) {

      // develop only check that modifiers isn't being updated needlessly
      if (process.env.NODE_ENV === "development") {
        if (
          this.props.modifiers !== prevProps.modifiers &&
          this.props.modifiers != null &&
          prevProps.modifiers != null &&
          shallowEqual(this.props.modifiers, prevProps.modifiers)
        ) {
          console.warn("'modifiers' prop reference updated even though all values appear the same.\nConsider memoizing the 'modifiers' object to avoid needless rendering.");
        }
      }

      if (this.popperInstance) {
        this.popperInstance.setOptions(this.getOptions())
      }
    }

    // A placement difference in state means popper determined a new placement
    // apart from the props value. By the time the popper element is rendered with
    // the new position Popper has already measured it, if the place change triggers
    // a size change it will result in a misaligned popper. So we schedule an update to be sure.
    if (prevState.placement !== this.state.placement) {
      this.update();
    }
  }

  componentWillUnmount() {
    setRef(this.props.innerRef, null)
    this.destroyPopperInstance();
  }

  render() {
    const { placement, isReferenceHidden, hasPopperEscaped } = this.state;

    return unwrapArray(this.props.children)({
      ref: this.setPopperNode,
      style: this.getPopperStyle(),
      placement,
      isReferenceHidden,
      hasPopperEscaped,
      update: this.update,
      arrowProps: {
        ref: this.setArrowNode,
        style: this.getArrowStyle(),
      },
    });
  }
}

export default function Popper({ referenceElement, ...props }: PopperProps) {
  return (
    <ManagerReferenceNodeContext.Consumer>
      {(referenceNode) => (
        <InnerPopper
          referenceElement={
            referenceElement !== undefined ? referenceElement : referenceNode
          }
          {...props}
        />
      )}
    </ManagerReferenceNodeContext.Consumer>
  );
}
