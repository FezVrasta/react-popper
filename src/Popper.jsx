import React, { Component, createElement } from 'react'
import PropTypes from 'prop-types'
import PopperJS from 'popper.js'
import isEqual from 'is-equal-shallow'

const noop = () => null

class Popper extends Component {
  static contextTypes = {
    popperManager: PropTypes.object.isRequired,
  }

  static childContextTypes = {
    popper: PropTypes.object.isRequired,
  }

  static propTypes = {
    component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    innerRef: PropTypes.func,
    placement: PropTypes.oneOf(PopperJS.placements),
    eventsEnabled: PropTypes.bool,
    modifiers: PropTypes.object,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }

  static defaultProps = {
    component: 'div',
    placement: 'bottom',
    eventsEnabled: true,
    modifiers: {},
  }

  state = {}

  subscribers = []

  getChildContext() {
    console.log('getChildContext', this)
    return {
      popper: {
        subscribe: this._subscribe,
        unsubscribe: this._unsubscribe,
        setArrowNode: this._setArrowNode,
      },
    }
  }

  // Add a subcomponent update fn that needs to get updated when we do
  _subscribe = fn => this.subscribers.push(fn)
  // Remove such a fn
  _unsubscribe = fn =>
    (this.subscribers = this.subscribers.filter(s => s !== fn))

  componentDidMount() {
    this._updatePopper()
  }

  componentDidUpdate(lastProps) {
    if (
      lastProps.placement !== this.props.placement ||
      lastProps.eventsEnabled !== this.props.eventsEnabled
    ) {
      this._updatePopper()
    }

    if (this._popper && lastProps.children !== this.props.children) {
      this._popper.scheduleUpdate()
    }
  }

  componentWillUnmount() {
    this._destroyPopper()
  }

  _setArrowNode = node => {
    this._arrowNode = node
  }

  _getTargetNode = () => {
    return this.context.popperManager.getTargetNode()
  }

  _updateFromData = data => {
    if (
      (this.state.data && !isEqual(data.offsets, this.state.data.offsets)) ||
      !this.state.data
    ) {
      this.setState({ data })
      this.subscribers.forEach(s => s(data))
    }
    return data
  }

  _updateStateModifier = {
    enabled: true,
    order: 900,
    fn: this._updateFromData,
  }

  _updatePopper() {
    this._destroyPopper()
    if (this._node) {
      this._createPopper()
    }
  }

  _createPopper() {
    const { placement, eventsEnabled } = this.props
    const modifiers = {
      ...this.props.modifiers,
      applyStyle: { enabled: false },
      updateState: this._updateStateModifier,
    }

    if (this._arrowNode) {
      modifiers.arrow = {
        element: this._arrowNode,
      }
    }

    this._popper = new PopperJS(this._getTargetNode(), this._node, {
      placement,
      eventsEnabled,
      modifiers,
    })

    // schedule an update to make sure everything gets positioned correct
    // after being instantiated
    this._popper.scheduleUpdate()
  }

  _destroyPopper() {
    if (this._popper) {
      this._popper.destroy()
    }
  }

  _getPopperStyle = () => {
    const { data } = this.state

    // If Popper isn't instantiated, hide the popperElement
    // to avoid flash of unstyled content
    if (!this._popper || !data) {
      return {
        position: 'absolute',
        pointerEvents: 'none',
        opacity: 0,
      }
    }

    const { top, left, position } = data.offsets.popper

    return {
      position,
      ...data.styles,
    }
  }

  _getPopperPlacement = () => {
    return !!this.state.data ? this.state.data.placement : undefined
  }

  render() {
    const {
      component,
      innerRef,
      placement,
      eventsEnabled,
      modifiers,
      children,
      ...restProps
    } = this.props

    const popperRef = node => {
      this._node = node
      if (typeof innerRef === 'function') {
        innerRef(node)
      }
    }
    const popperStyle = this._getPopperStyle()
    const popperPlacement = this._getPopperPlacement()

    if (typeof children === 'function') {
      const popperProps = {
        ref: popperRef,
        style: popperStyle,
        ['data-placement']: popperPlacement,
      }
      return children({
        popperProps,
        restProps,
        scheduleUpdate: this._popper && this._popper.scheduleUpdate,
      })
    }

    const componentProps = {
      ...restProps,
      style: {
        ...restProps.style,
        ...popperStyle,
      },
      'data-placement': popperPlacement,
    }

    if (typeof component === 'string') {
      componentProps.ref = popperRef
    } else {
      componentProps.innerRef = popperRef
    }

    return createElement(component, componentProps, children)
  }
}

export default Popper
