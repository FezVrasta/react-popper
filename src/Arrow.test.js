// @flow
import React from 'react';
import { mount } from 'enzyme';

// Public API
import { Arrow } from '.';

// Private API
import { ArrowNodeContext } from './Popper';

describe('Arrow component', () => {
  it('renders the expected markup', () => {
    const wrapper = mount(
      <Arrow>
        {({ getArrowRef, style }) => <div ref={getArrowRef} style={style} />}
      </Arrow>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('consumes the ArrowNodeContext from Popper', () => {
    const setArrowNode = jest.fn();
    const arrowStyle = { top: 10 };

    // HACK: wrapping DIV needed to make Enzyme happy for now
    const wrapper = mount(
      <div>
        <ArrowNodeContext.Provider
          value={{
            setArrowNode,
            arrowNode: null,
            arrowStyle,
          }}
        >
          <Arrow>
            {({ getArrowRef, style }) => (
              <div ref={getArrowRef} style={style} />
            )}
          </Arrow>
        </ArrowNodeContext.Provider>
      </div>
    );
    expect(setArrowNode).toHaveBeenCalled();
    expect(wrapper.children()).toMatchSnapshot();
  });
});
