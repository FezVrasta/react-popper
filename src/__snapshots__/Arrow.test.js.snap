// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow component consumes the ArrowNodeContext from Popper 1`] = `
<Component>
  <div
    style={
      Object {
        "top": 10,
      }
    }
  />
</Component>
`;

exports[`Arrow component renders the expected markup 1`] = `
<Component>
  <div
    style={Object {}}
  />
</Component>
`;
